File: 2021/archived/Day06/Day6.fs
module Day6

open System.IO

let init days =
    let fishes = [| for _ in 0..8 do 0UL |]
    let addFish fishes fish =
        fishes
        |> Array.mapi (fun i f -> if i = fish then f + 1UL else f)

    days
    |> Array.fold addFish fishes

let nextDay (fishes: uint64 array) =
    let additionalSix, nextEight = fishes.[0], fishes.[0]

    Array.append fishes [| nextEight |]
    |> Array.pairwise
    |> Array.mapi (fun i (_, next) ->
        if i <> 6 then
            next
        else
            next + additionalSix)

let solve left fishes =
    let rec loop left fishes =
        if left = 0 then
            fishes
        else
            fishes |> nextDay |> loop (left - 1)

    loop left fishes |> Array.sum

let input =
    (File.ReadAllLines "input.txt" |> Array.head)
        .Split [| ',' |]
    |> Array.map int
    |> init

solve 80 input |> printfn "First part: %d"
solve 256 input |> printfn "Second part: %d"

File: 2021/archived/Day06/Day6.scala
import scala.io.Source
import scala.util.chaining._
import scala.language.implicitConversions
import scala.annotation.tailrec

@main def day6 =
    def init(days: Array[Int]) =
        val fishes = for i <- 0 to 8 yield 0

        days
            .foldLeft(fishes)((fs, f) =>
                val curr = fs(f)
                fs.updated(f, curr + 1)
            )
            .map(_.toLong)
            .toVector

    def nextDay(fishes: Vector[Long]) =
        val additionalSix, nextEight = fishes(0)

        (fishes :+ nextEight)
            .sliding(2)
            .zipWithIndex
            .map((pair, day) =>
                val Vector(_, next) = pair
                if day == 6 then next + additionalSix
                else next
            )
            .toVector

    def solve(left: Int, fishes: Vector[Long]) =
        @tailrec
        def loop(left: Int, fishes: Vector[Long]): Vector[Long] =
            if left == 0 then fishes
            else loop(left - 1, nextDay(fishes))

        loop(left, fishes).sum

    val input = Source
        .fromFile("input.txt")
        .getLines
        .next
        .split(',')
        .map(_.toInt)
        .pipe(init)

    println(s"First part: ${solve(80, input)}")
    println(s"Second part: ${solve(256, input)}")
